File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string>sett(wordList.begin(), wordList.end());
        queue<pair<string, int>> q; //word, steps
        q.push({beginWord, 1});
        sett.erase(beginWord);

        while(!q.empty()){
            string word = q.front().first;
            int steps = q.front().second;
            q.pop();
            if(word == endWord) return steps;
            for(int i = 0; i < word.size(); i++){
                char original = word[i];
                for(char c = 'a'; c <='z'; c++){
                    word[i] = c;

                    if(sett.find(word) != sett.end()){
                        q.push({word, steps + 1});
                        sett.erase(word);
                    }
                }
                word[i] = original;
            }
        }
        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        sett = set(wordList)
        q = deque()
        q.append((beginWord, 1))
        sett.discard(beginWord)   
        # sett.remove(beginWord) #leads to an error if element not present in set   

        while q:
            word, steps = q.popleft()
            if word == endWord:
                return steps
            
            for i in range(len(word)):
                for c in string.ascii_lowercase:
                    new_word = word[:i] + c + word[i+1:]
                    if new_word in sett:
                        q.append((new_word, steps+1))
                        sett.remove(new_word)
        return 0
 